### PR TITLE
[RFC] Use 6to5 and take advantage of rest transforms (avoids leaking arguments)

### PR DESCRIPTION
**Pull request by camspiers**
The is mostly a request for comment as I understand moving to 6to5 might have other implications. But tests all pass and performance improves in some cases. Thoughts?

The rest params transform looks a lot like your `arrCopy` (minus the leaking of args because it is inline):

```
for (var _len4 = arguments.length, values = Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
    values[_key4] = arguments[_key4];
}
```

```
Running "stats" task
  Raw: 142,546 bytes         +4,050 bytes
  Min:  58,247 bytes 59.13%  +2,215 bytes
  Zip:  15,823 bytes 88.89%    +122 bytes
```

```
List > builds from array of 2
  Old:   691,602   699,059   706,679 ops/sec
  New:   808,152   815,852   823,701 ops/sec
  compare: 1 1
  diff: 16.7%
  rme: 1.01%
List > builds from array of 8
  Old:   640,498   645,867   651,326 ops/sec
  New:   740,372   746,932   753,609 ops/sec
  compare: 1 1
  diff: 15.64%
  rme: 0.86%
List > builds from array of 32
  Old:   254,254   256,676   259,145 ops/sec
  New:   272,485   274,272   276,083 ops/sec
  compare: 1 1
  diff: 6.85%
  rme: 0.81%
List > builds from array of 1024
  Old:     9,785     9,870     9,957 ops/sec
  New:     9,848     9,930    10,012 ops/sec
  compare: 1 1
  diff: 0.6%
  rme: 0.85%
List > pushes into 2 times
  Old:   839,622   844,886   850,217 ops/sec
  New:   877,473   885,676   894,033 ops/sec
  compare: 1 1
  diff: 4.82%
  rme: 0.79%
List > pushes into 8 times
  Old:   189,652   191,290   192,956 ops/sec
  New:   191,637   196,253   201,097 ops/sec
  compare: 1 1
  diff: 2.59%
  rme: 1.8%
List > pushes into 32 times
  Old:    45,151    45,543    45,941 ops/sec
  New:    45,781    46,111    46,446 ops/sec
  compare: 1 1
  diff: 1.24%
  rme: 0.79%
List > pushes into 1024 times
  Old:     1,394     1,406     1,419 ops/sec
  New:     1,475     1,487     1,499 ops/sec
  compare: 1 1
  diff: 5.7%
  rme: 0.84%
List > pushes into transient 2 times
  Old: 1,328,763 1,341,117 1,353,704 ops/sec
  New: 1,478,902 1,489,587 1,500,428 ops/sec
  compare: 1 1
  diff: 11.07%
  rme: 0.83%
List > pushes into transient 8 times
  Old:   525,532   530,718   536,007 ops/sec
  New:   534,015   538,823   543,719 ops/sec
  compare: 1 1
  diff: 1.52%
  rme: 0.94%
List > pushes into transient 32 times
  Old:   162,114   163,528   164,967 ops/sec
  New:   162,803   164,121   165,460 ops/sec
  compare: 0 0
  diff: 0.36%
  rme: 0.84%
List > pushes into transient 1024 times
  Old:     5,002     5,130     5,265 ops/sec
  New:     5,281     5,329     5,379 ops/sec
  compare: 1 1
  diff: 3.88%
  rme: 1.92%
Map builds from an object of 2
  Old:   531,265   536,883   542,622 ops/sec
  New:   480,687   485,541   490,495 ops/sec
  compare: 1 1
  diff: -9.57%
  rme: 1.03%
Map builds from an object of 8
  Old:   105,747   106,688   107,646 ops/sec
  New:   101,170   102,090   103,026 ops/sec
  compare: 1 1
  diff: -4.32%
  rme: 0.89%
Map builds from an object of 32
  Old:    37,285    37,653    38,028 ops/sec
  New:    37,252    37,552    37,856 ops/sec
  compare: 0 0
  diff: -0.27%
  rme: 0.9%
Map builds from an object of 1024
  Old:     1,836     1,852     1,868 ops/sec
  New:     1,758     1,801     1,845 ops/sec
  compare: 1 1
  diff: -2.76%
  rme: 1.81%
all done
```

### CODE REVIEW

**Comment by leebyron** [comment]
Yes! I actually have an almost identical change sitting in a branch on my local machine to do this, and I've been building out the perf tests in order to measure exactly the question you're asking about implications.

> The rest params transform looks a lot like your arrCopy (minus the leaking of args because it is inline):

Yep! I added that change to 6to5!

So I'm all for 6to5, but just need more time to understand the implications first.

File: src/Hash.js [hunk 1]
```diff
@@ -111,8 +111,8 @@ function hashJSObj(obj) {
     // we'll hijack one of the less-used non-enumerable properties to
     // save our hash on it. Since this is a function it will not show up in
     // `JSON.stringify` which is what we want.
-    obj.propertyIsEnumerable = function() {
-      return this.constructor.prototype.propertyIsEnumerable.apply(this, arguments);
+    obj.propertyIsEnumerable = function(...values) {
+      return this.constructor.prototype.propertyIsEnumerable.apply(this, values);
```
Comment (leebyron) [inline src/Hash.js]: FYI: JS VMs will interpret `fn.apply(this, arguments)` as not leaking the arguments object but instead forward the call on to the new target.

Comment (camspiers) [inline src/Hash.js]: Interesting, thanks for the information!

**Comment by sebmck** [comment]
Would be interested to find out the cases that 6to5 has actually made slower. Looks like definite areas for improvement.

File: src/IterableImpl.js [hunk 2]
```diff
@@ -233,9 +233,9 @@ mixin(Iterable, {
     return reduction;
   },
 
-  reduceRight(reducer, initialReduction, context) {
+  reduceRight(...values) {
     var reversed = this.toKeyedSeq().reverse();
-    return reversed.reduce.apply(reversed, arguments);
+    return reversed.reduce.apply(reversed, values);
```
Comment (sebmck) [inline src/IterableImpl.js]: These could be simplified further to just:

```javascript
return reversed.reduce(...arguments);
```

Comment (camspiers) [inline src/IterableImpl.js]: Wow, I didn't know that.

**Comment by camspiers** [comment]
@sebmck just speculating, but I think that it is the conversion from the array-like `arguments` to array. This is good as it avoids leaking args, but it isn't without cost.

https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments

**Comment by sebmck** [comment]
@camspiers As @leebyron said engines optimise for `fn.apply(this, arguments);` and doesn't leak.

**Comment by camspiers** [comment]
A lot of the changes don't use `apply`, but yes you are definitely right that some of the changes are unnecessary (and should be changed). Thanks heaps for the information :+1:

**Comment by sebmck** [comment]
@camspiers Ah right sorry, I just skimmed over the changes. Looks good though! Let me know if you find any transformation output that can be optimised. Might be worth taking a look at [loose mode](http://6to5.org/docs/usage/loose/) if immutable uses any of the features specified on that page.

**Comment by facebook-github-bot-6** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Pull request closed without merging**